content/blog/2022-04-26-Superoptimizer.md

```diff
@@ -0,0 +1,83 @@
+
+Denali operates faster than the original Superoptimizer. Denali is able to generate sequences of instructions ranging up to 31 instructions in length, compared to around 13 for Superoptimizer. Denali also only generates semantically equivalent programs, as compared to Superoptimizer, which might require the user to prove that the generated program is equivalent to the original program. However, the programs that Denali generates are not necessarily minimally sized. In particular, if the minimal size program is not generatable via the rewrite rules that Denali uses, then it cannot be an output of Denali. As a result, there is a tradeoff between speed of the system, versus the ability to completely search the program space to find an optimal solution. Depending on the application, such a tradeoff might be acceptable, particularly when speed of compilation matters more, or when the number of instructions to optimize is larger in size. 
+
+Finally, to improve upon the speed of the Superoptimizer, the STOKE project uses a different search method [^4]. In this project, stochastic search is employed to first search the program space for enumerated programs, and then an optimized program is chosen from these subspaces. This approach is able to scale up to larger programs, using a shorter amount of time to compile, compared to the original Superoptimizer. The authors of STOKE even find that on certain benchmarks, optimized output programs can surpass -O3 and hand optimized programs. Once again, there is a tradeoff where the most minimal program might not be found, as compared to Superoptimizer. However, this tradeoff may not be that important, because STOKE’s ability to beat hand optimized programs, particularly on larger sized programs, allows the ability to bring superoptimization to a larger space of applications, which can benefit more programmers across the spectrum of programming. 
+
+## Conclusion
+
+Massalin’s work is seminal, because it is the first to consider truly optimal programs, rather than incremental improvements that constitute most compiler optimizations. Being able to find optimal real world programs such as the signum function was a major success of Massalin’s work. However, Superoptimizer has several downsides such as the time spent searching for programs, and the inability to guarantee superoptimized programs correct by construction. These downsides are dealt with in other works that build upon ideas found in Superoptimizer. All together, the Superoptimizer provided a practical approach, yet novel, to solve an important problem. Today, the techniques pioneered in this paper continue to be used and improved upon to provide new solutions. 
```

Comment:
This is a fine conclusion, but it could be enhanced even further by playing up the fact that the idea seems impractical on its face. A main takeaway in the paper is that, despite this "obvious" impracticality in general, it _can actually work_ in some situations. And that's what has been so inspiring to subsequent work.

Reply:
I fixed up the conclusion to focus on the point that you made about how the impractical solution can still lead to impressive results when optimization certain programs.